semgrep-core/src/cli-bridge/bridge_design.txt

```diff
@@ -0,0 +1,369 @@
+
+
+Summary
+-------
+
+The 'semgrep_bridge_python' module acts as a bridge from the Semgrep
+command line interface (CLI), which is implemented in Python, to the
+Semgrep analysis core, which is written in OCaml.  It allows the
+analysis to be run from within the calling Python process, as a Python
+module.  Furthermore, it lets the CLI and analysis communicate entirely
+in memory rather than using any files on disk.
+
+
+Background
+----------
+
+Some users of Semgrep want to be able to run it within the Snowflake
+hosted database system.  Snowflake allows users to extend the system by
+creating user-defined functions (UDFs) in several languages, including
+Python, but UDFs have restricted access permissions, both for security
+reasons and to avoid dependence on execution hardware.  It is our
+current understanding that the primary restriction is the inability
+to access any file system.
+
+Separately, there are potential performance advantages to running the
+Semgrep analysis without launching another process.
+
+That said, the ability to run the analysis as a separate process should
+not be removed, since it is useful for development and debugging.
+Consequently, the bridge is an optional component, integrated in a way
+that is minimally intrusive on both sides.
+
+
+Bridge design
+-------------
+
+See diagram: bridge_design.ded.png.  (Editable with
+https://github.com/smcpeak/ded)
+
+The bridge consists of two shared libraries:
+
+  * semgrep_bridge_python.so: Thin layer with Python bindings,
+    implemented in bridge_py.c.
+
+  * semgrep_bridge_core.so: OCaml bindings (bridge_ml.c) and the
+    analysis core (in OCaml).
+
+To provide analysis inputs, the bridge initially forwards the list of
+command line arguments.  Some of these arguments are the names of input
+files that would normally be read from disk.  So, the bridge also
+provides the analysis a callback to get the contents of a named file.
+In stand-alone executable operation, that callback reads from disk as
+usual (using OCaml file I/O).  In bridge module operation, it relays the
+request back to the Python code, which serves requests using Python file
+I/O.
+
+When the analysis finishes, it returns control to the bridge, which in
+turn returns control to the CLI.  If the analysis terminates with an
+uncaught exception, that is converted to a string and delivered to the
+CLI in that form.
+
+
+Finding shared libraries
+------------------------
+
+In the Python CLI, the existing logic for locating semgrep-core has been
+adapted to also locate semgrep_bridge_python.so.  That file then looks
+for semgrep_bridge_core.so in whatever directory it was loaded from by
+taking advantage of the "-Wl,-rpath=$ORIGIN" linker option.
+
+
+Analysis runs in a forked child
+-------------------------------
+
+Prior to invoking the bridge module, core_runner.py forks a child
+process in which to invoke the bridge.  (Note: This is fork without
```

Comment:
See last paragraph in that section: "It is unknown whether Snowflake UDFs are allowed to invoke fork."  :)  Basically, this fork is needed for things to work in normal usage, and I've got a contingency plan in case it is not allowed in a UDF.